benchmark/single-source/CharacterProperties.swift.gyb

```diff
@@ -0,0 +1,108 @@
+//===--- StringMatch.swift ------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+% # Ignore the following warning. This _is_ the correct file to edit.
+////////////////////////////////////////////////////////////////////////////////
+// WARNING: This file is manually generated from .gyb template and should not
+// be directly modified. Instead, make changes to StringWalk.swift.gyb and run
+// scripts/generate_harness/generate_harness.py to regenerate this file.
+////////////////////////////////////////////////////////////////////////////////
+
+import TestsUtils
+import Foundation
+
+public let CharacterPropertiesFetch = BenchmarkInfo(
+  name: "CharacterPropertiesFetch",
+  runFunction: run_CharacterPropertiesFetch,
+  tags: [.validation, .api, .String])
+
+public let CharacterPropertiesStashed = BenchmarkInfo(
+  name: "CharacterPropertiesStashed",
+  runFunction: run_CharacterPropertiesStashed,
+  tags: [.validation, .api, .String])
+
+
+extension Character {
+  var firstScalar: UnicodeScalar { return unicodeScalars.first! }
+}
+
+% Properties = { "Alphanumeric": "alphanumerics", "Capitalized": "capitalizedLetters", \
+%              "Control": "controlCharacters", \
+%              "Decimal": "decimalDigits", \
+%              "Letter": "letters", \
+%              "Lowercase": "lowercaseLetters", \
+%              "Uppercase": "uppercaseLetters", \
+%              "Newline": "newlines", \
+%              "Whitespace": "whitespaces", \
+%              "Punctuation": "punctuationCharacters" \
+%              }
+
+// Fetch the CharacterSet for every call
+% for Property, Set in Properties.items():
+func is${Property}(_ c: Character) -> Bool { 
+  return CharacterSet.${Set}.contains(c.firstScalar)
+}
+% end
+
+// Stash the set
+% for Property, Set in Properties.items():
+let ${Set} = CharacterSet.${Set}
+func is${Property}Stashed(_ c: Character) -> Bool { 
+  return ${Set}.contains(c.firstScalar)
+}
+% end
+
+// Compute on the fly
+//
+// TODO: If UnicodeScalars ever exposes category, etc., implement the others!
+func isNewlineComputed(_ c: Character) -> Bool {
```

Comment:
We only have an implementation of isNewline because CharacterSet's newlines are not defined in terms of runtime-dependent assignment of Unicode general categories, but are fixed.

A `run_CharacterPropertiesComputed` that only executes on newlines wouldn't be comparable to others, so I'm not sure it gives a useful signal.

Reply:
I'm open to suggestions :-)